### PR TITLE
Update to support Django versions < 5.0

### PR DESCRIPTION
**Pull request by Afani97**
- Update django to latest 4.2 version
- Set a max version allowed for `smpplib` until a bug fix PR has been merged, noted in the toml file. 
- Update pre-commit file to include `django-upgrade`

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -14,9 +14,10 @@ include = [ { path = "tests", format = "sdist" } ]
 [tool.poetry.dependencies]
 python = ">=3.8"
 RapidSMS = ">=2.0"
```
Comment (Afani97) [inline pyproject.toml:16]: `RapidSMS` might need to be upgraded, it seems to be crashing this build: https://github.com/caktus/rapidsms-smpp-gateway/actions/runs/5581453954/jobs/10199938046?pr=16#step:6:67

**Review by ronardcaktus — COMMENTED** [review]
Looks good. Maybe `changes.md` should be updated with release notes?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -9,6 +9,11 @@ repos:
         language: system
         pass_filenames: false
         types: ["file", "python"]
+  - repo: https://github.com/adamchainz/django-upgrade
+    rev: 1.14.0
```
[the remark this answers is not in the document]
Comment (Afani97) [reply .pre-commit-config.yaml:13]: That sounds good to me, I'll see what I can do.

**Comment by ronardcaktus** [comment]
Github actions is failing with the following error:
```
  File "/home/runner/.cache/pypoetry/virtualenvs/smpp-gateway-28Xge_hh-py3.10/lib/python3.10/site-packages/rapidsms/templatetags/paginator_tags.py", line 19, in <module>
    if "django.template.context_processors.request" not in settings.TEMPLATES[0]['OPTIONS']['context_processors']:
IndexError: list index out of range
```

However, `settings.TEMPLATES[0]["OPTIONS"]["context_processors"]` is in the code. I did some local testing:
```
(Pdb) settings.TEMPLATES[0]["OPTIONS"]["context_processors"]
['django.contrib.auth.context_processors.auth', 'django.template.context_processors.debug', 
'django.template.context_processors.i18n', 'django.template.context_processors.media', 'django.template.context_processors.static', 'django.template.context_processors.tz', 'django.contrib.messages.context_processors.messages', 'django.template.context_processors.request']

(Pdb) "django.template.context_processors.request" in settings.TEMPLATES[0]["OPTIONS"]["context_processors"]
True
```

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -9,7 +9,7 @@ jobs:
 
     services:
       postgres:
-        image: postgres:10.18
+        image: postgres:12.16
```
Comment (Afani97) [inline .github/workflows/ci.yaml:12]: Ran into this:
```
  File "/home/runner/.cache/pypoetry/virtualenvs/smpp-gateway-28Xge_hh-py3.8/lib/python3.8/site-packages/django/db/backends/base/base.py", line 239, in init_connection_state
    self.check_database_version_supported()
  File "/home/runner/.cache/pypoetry/virtualenvs/smpp-gateway-28Xge_hh-py3.8/lib/python3.8/site-packages/django/db/backends/base/base.py", line 214, in check_database_version_supported
    raise NotSupportedError(
django.db.utils.NotSupportedError: PostgreSQL 12 or later is required (found 10.18).
```

**Pull request merged**